feat: Add Dynamic Versioning SCM plugin

jimisola opened the pull request:
Adds dynamic versioning based on SCM tag (git supported for now). This plugin goal differs from other Versions Maven Plugin goals since it _does not_ change the pom.xml but simply updates project.version in-memory during execution.

Closes: https://github.com/mojohaus/versions/issues/1071

jimisola commented:
@slawekjaranowski I suddenly keep getting: 
 
```
$mvn verify -P run-its -pl versions-maven-plugin -DskipTests -Dinvoker.test=it-ci-friendly-versions-001

 [INFO] --- animal-sniffer:1.23:check (check-java18) @ versions-maven-plugin ---
[INFO] Checking unresolved references to org.codehaus.mojo.signature:java18:1.0
[ERROR] /home/jimisola/cloud/dev/clones/jimisola/github/versions/versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsReport.java: Field ruleSet: Undefined reference: RuleSet
[ERROR] /home/jimisola/cloud/dev/clones/jimisola/github/versions/versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractVersionsUpdaterMojo.java: Field ruleSet: Undefined reference: RuleSet
```

Not sure how this build is setup but tried rebuilding all the modules from root with a simple `mvn clean install`.
However, the RuleSet is missing. Must a do something separate to create the RuleSet class using MODELLO?

lfvjimisola commented:
Fully implemented and it pass with:

`mvn verify -P run-its -pl versions-maven-plugin -DskipTests -Dinvoker.test=it-ci-friendly-versions-*`

Will squash commits later and change to ready.

@slawekjaranowski How does it look?

jimisola commented:
@slawekjaranowski I managed to do something for it not work. Can't figure out what it is. I changed the name of plugin class. The project version seems to be set correctly, but somehow the build artifacts are not using the set version.

I tried with phases validate and initialize with the same result. 

Do you have any pointer as to what could be wrong? I would really to avoid having to create an extension.

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/DynamicVersioningSCMPlugin.java:
```diff
@@ -0,0 +1,259 @@
+     *         parameter {@code useVersion}
+     */
+    protected Optional<String> getHardcodedVersion() {
+        String version = System.getenv(DV_SCM_USE_VERSION_ENV_VAR_NAME);
```
It is not usual to use environments for plugin configuration.

Why you need it?

jimisola replied:
It is for bypassing the whole SCM functionality. Used e.g. when testing pipelines.

From a GitLab CI/CD Components perspective we could set an environment variable when testing that will add `-DuseVersion=true` to the Maven options.

However, not using an environment variable adds "work" and does not feel as reliable.
This since using `useVersion` from the command line in a Maven build with many plugins there is a risk that `useVersion` is used in other plugins. Whereas the environment variable is more or less guaranteed to be unique.

And option would be to change the parameters so that they are more unique, e.g. like Spring Java Format does with `spring-javaformat.apply.skip`.

The parameter `useVersion` shall not be set in the POM configuration as it main purpose is for testing.

I can remove if it is a deal-breaker but  I think we should think about the naming of the parameters then.

lfvjimisola replied:
Removed.

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/DynamicVersioningSCMPlugin.java:
```diff
@@ -0,0 +1,259 @@
+        if (version != null) {
+            return Optional.of(version);
+        } else if (useVersion != null) {
+            return Optional.of(useVersion);
```
`ofNullable`

jimisola replied:
Like this?

```java
    return Optional.ofNullable(System.getenv(DV_SCM_USE_VERSION_ENV_VAR_NAME))
                   .or(() -> Optional.ofNullable(useVersion));
```

lfvjimisola replied:
Moved/removed.

slawekjaranowski commented on versions-maven-plugin/src/it/it-dynamic-versioning-scm-tag-0-commits-0-tags/invoker.properties:
```diff
@@ -0,0 +1,2 @@
+invoker.goals=${project.groupId}:${project.artifactId}:${project.version}:set-version-from-scm-tag
+invoker.buildResult = failure
```
missing empty line at the and of file

jimisola replied:
For this file only or also the others with :no_entry: ?

My settings in VS Code seems to cause this. Does not `mvn spotless:apply` format this correctly?

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/DynamicVersioningSCMPlugin.java:
```diff
@@ -0,0 +1,259 @@
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
+
+/**
+ * Maven plugin that uses SCM/VCS to enable dynamic versioning based on your
+ * version control system.
+ *
+ * @author Jimisola Laursen
+ * @since 2.17.0
+ */
+@Mojo(name = "set-version-from-scm-tag", defaultPhase = LifecyclePhase.INITIALIZE)
```
This goal only discover version based on SCM and set project property with it ... 

Most of goals of this plugin manipulate project pom to update it according to goal.

We need to think about name

jimisola replied:
Ok. I'm all ears about the name of the goal as long as the functionality is there. What are your thoughts about the name?

lfvjimisola replied:
@slawekjaranowski Do you have a suggestion?

jimisola replied:
Brainstorming here, but the key elements to get in the goal are at least:

1. updates project.version in-memory (dynamic)
2. sets version based on tag in scm

Is there another word that set that would indicate that the pom.xml itself is left untouched?

The best I can come up with now are:

* use-dynamic-version-from-scm: execution will use but not set version
* impersonate-dynamic-version-from-scm: sadly, inversionate is not a word ;)

slawekjaranowski replied:
`use-dynamic-version-from-scm` look ok for me 😄

jimisola replied:
Ok. Let's go with use-dynamic-version-from-scm then.

lfvjimisola commented:
@slawekjaranowski Would it be possible for us to get this moving so that it can be merged to main before the next release?

slawekjaranowski commented:
please:
 - check output of build
 - rebase with current master branch

jimisola commented:
> please:
> 
> * check output of build

Not sure that I'm following. After rebase, I ran `mvn verify -P run-its -pl versions-maven-plugin` with BUILD SUCCESS - if that was it. Can you please explain further if that is not what you where asking?

> * rebase with current master branch

Should be done now.

slawekjaranowski commented:
> > please:
> > 
> > * check output of build
> 
> Not sure that I'm following. After rebase, I ran `mvn verify -P run-its -pl versions-maven-plugin` with BUILD SUCCESS - if that was it. Can you please explain further if that is not what you where asking?

Pleas look at https://github.com/mojohaus/versions/actions/runs/9527983742/job/26266186230

Please also rebuild whole project `mvn verify -P run-its` to be sure

lfvjimisola commented:
> > > please:
> > > 
> > > * check output of build
> > 
> > 
> > Not sure that I'm following. After rebase, I ran `mvn verify -P run-its -pl versions-maven-plugin` with BUILD SUCCESS - if that was it. Can you please explain further if that is not what you where asking?
> 
> Pleas look at https://github.com/mojohaus/versions/actions/runs/9527983742/job/26266186230
> 
> Please also rebuild whole project `mvn verify -P run-its` to be sure

I've fixed the spotless issue and removed a test that should have been removed. Everything passed.

After I resolved conflicts, i.e. merged in changes from master using GitHib GUI, `mvn verify -P run-its` fails.
There seems to be an issue with unresolved dependencies. I realized now that perhaps I should have rebased instead as before.

Would you mind having a look? 

```bash
[ERROR] Tests run: 3, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 0.174 s <<< FAILURE! -- in org.codehaus.mojo.versions.UpdatePropertyMojoTest
[ERROR] org.codehaus.mojo.versions.UpdatePropertyMojoTest.testAllowMajorUpdates -- Time elapsed: 0.059 s <<< ERROR!
java.lang.NoSuchMethodError: 'java.lang.String org.codehaus.mojo.versions.UpdatePropertyMojo.access$0(org.codehaus.mojo.versions.UpdatePropertyMojo)'
        at org.codehaus.mojo.versions.UpdatePropertyMojo$1.<init>(UpdatePropertyMojo.java:100)
        at org.codehaus.mojo.versions.UpdatePropertyMojo.update(UpdatePropertyMojo.java:96)
        at org.codehaus.mojo.versions.AbstractVersionsUpdaterMojo.process(AbstractVersionsUpdaterMojo.java:316)
        at org.codehaus.mojo.versions.AbstractVersionsUpdaterMojo.execute(AbstractVersionsUpdaterMojo.java:270)
        at org.codehaus.mojo.versions.UpdatePropertyMojoTest.testAllowMajorUpdates(UpdatePropertyMojoTest.java:40)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at junit.framework.TestCase.runTest(TestCase.java:177)
        at junit.framework.TestCase.runBare(TestCase.java:142)
        at junit.framework.TestResult$1.protect(TestResult.java:122)
        at junit.framework.TestResult.runProtected(TestResult.java:142)
        at junit.framework.TestResult.run(TestResult.java:125)
        at junit.framework.TestCase.run(TestCase.java:130)
        at junit.framework.TestSuite.runTest(TestSuite.java:241)
        at junit.framework.TestSuite.run(TestSuite.java:236)

[ERROR] org.codehaus.mojo.versions.UpdatePropertyMojoTest.testAllowIncrementalUpdates -- Time elapsed: 0.057 s <<< ERROR!
java.lang.NoSuchMethodError: 'java.lang.String org.codehaus.mojo.versions.UpdatePropertyMojo.access$0(org.codehaus.mojo.versions.UpdatePropertyMojo)'
        at org.codehaus.mojo.versions.UpdatePropertyMojo$1.<init>(UpdatePropertyMojo.java:100)
        at org.codehaus.mojo.versions.UpdatePropertyMojo.update(UpdatePropertyMojo.java:96)
        at org.codehaus.mojo.versions.AbstractVersionsUpdaterMojo.process(AbstractVersionsUpdaterMojo.java:316)
        at org.codehaus.mojo.versions.AbstractVersionsUpdaterMojo.execute(AbstractVersionsUpdaterMojo.java:270)
        at org.codehaus.mojo.versions.UpdatePropertyMojoTest.testAllowIncrementalUpdates(UpdatePropertyMojoTest.java:73)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at junit.framework.TestCase.runTest(TestCase.java:177)
        at junit.framework.TestCase.runBare(TestCase.java:142)
        at junit.framework.TestResult$1.protect(TestResult.java:122)
        at junit.framework.TestResult.runProtected(TestResult.java:142)
        at junit.framework.TestResult.run(TestResult.java:125)
        at junit.framework.TestCase.run(TestCase.java:130)
        at junit.framework.TestSuite.runTest(TestSuite.java:241)
        at junit.framework.TestSuite.run(TestSuite.java:236)

[ERROR] org.codehaus.mojo.versions.UpdatePropertyMojoTest.testAllowMinorUpdates -- Time elapsed: 0.056 s <<< ERROR!
java.lang.NoSuchMethodError: 'java.lang.String org.codehaus.mojo.versions.UpdatePropertyMojo.access$0(org.codehaus.mojo.versions.UpdatePropertyMojo)'
        at org.codehaus.mojo.versions.UpdatePropertyMojo$1.<init>(UpdatePropertyMojo.java:100)
        at org.codehaus.mojo.versions.UpdatePropertyMojo.update(UpdatePropertyMojo.java:96)
        at org.codehaus.mojo.versions.AbstractVersionsUpdaterMojo.process(AbstractVersionsUpdaterMojo.java:316)
        at org.codehaus.mojo.versions.AbstractVersionsUpdaterMojo.execute(AbstractVersionsUpdaterMojo.java:270)
        at org.codehaus.mojo.versions.UpdatePropertyMojoTest.testAllowMinorUpdates(UpdatePropertyMojoTest.java:56)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at junit.framework.TestCase.runTest(TestCase.java:177)
        at junit.framework.TestCase.runBare(TestCase.java:142)
        at junit.framework.TestResult$1.protect(TestResult.java:122)
        at junit.framework.TestResult.runProtected(TestResult.java:142)
        at junit.framework.TestResult.run(TestResult.java:125)
        at junit.framework.TestCase.run(TestCase.java:130)
        at junit.framework.TestSuite.runTest(TestSuite.java:241)
        at junit.framework.TestSuite.run(TestSuite.java:236)

[INFO] Running org.codehaus.mojo.versions.MinimalMavenBuildVersionFinderTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 s -- in org.codehaus.mojo.versions.MinimalMavenBuildVersionFinderTest
[INFO] Running org.codehaus.mojo.versions.SetMojoTest
[INFO] Tests run: 16, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.041 s -- in org.codehaus.mojo.versions.SetMojoTest
[INFO] Running org.codehaus.mojo.versions.UpdatePropertiesMojoTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.295 s -- in org.codehaus.mojo.versions.UpdatePropertiesMojoTest
[INFO] Running org.codehaus.mojo.versions.PropertyUpdatesReportMojoTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.149 s -- in org.codehaus.mojo.versions.PropertyUpdatesReportMojoTest
[INFO] Running org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest
[ERROR] Tests run: 4, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 0.005 s <<< FAILURE! -- in org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest
[ERROR] org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.testIncludesMakesSetNonEmpty -- Time elapsed: 0.002 s <<< ERROR!
java.lang.NoSuchMethodError: 'java.nio.file.Path org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.access$0(org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest)'
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest$2.<init>(DisplayExtensionUpdatesMojoTest.java:88)
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.setUp(DisplayExtensionUpdatesMojoTest.java:84)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)

[ERROR] org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.testIncludesExcludesMakesSetEmpty -- Time elapsed: 0.001 s <<< ERROR!
java.lang.NoSuchMethodError: 'java.nio.file.Path org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.access$0(org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest)'
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest$2.<init>(DisplayExtensionUpdatesMojoTest.java:88)
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.setUp(DisplayExtensionUpdatesMojoTest.java:84)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)

[ERROR] org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.testNoBuildExists -- Time elapsed: 0.001 s <<< ERROR!
java.lang.NoSuchMethodError: 'java.nio.file.Path org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.access$0(org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest)'
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest$2.<init>(DisplayExtensionUpdatesMojoTest.java:88)
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.setUp(DisplayExtensionUpdatesMojoTest.java:84)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)

[ERROR] org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.testIncludesMakesSetEmpty -- Time elapsed: 0 s <<< ERROR!
java.lang.NoSuchMethodError: 'java.nio.file.Path org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.access$0(org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest)'
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest$2.<init>(DisplayExtensionUpdatesMojoTest.java:88)
        at org.codehaus.mojo.versions.DisplayExtensionUpdatesMojoTest.setUp(DisplayExtensionUpdatesMojoTest.java:84)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)

[INFO] Running org.codehaus.mojo.versions.DependencyUpdatesXmlRendererTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0 s <<< FAILURE! -- in org.codehaus.mojo.versions.DependencyUpdatesXmlRendererTest
[ERROR] org.codehaus.mojo.versions.DependencyUpdatesXmlRendererTest.testReportGeneration -- Time elapsed: 0 s <<< ERROR!
java.lang.Error: 
Unresolved compilation problems: 
        The import org.codehaus.mojo.versions.reporting.model.DependencyInfo cannot be resolved
        The import org.codehaus.mojo.versions.reporting.model.DependencyReportSummary cannot be resolved
        The import org.codehaus.mojo.versions.reporting.model.DependencyUpdatesReport cannot be resolved
        The import org.codehaus.mojo.versions.reporting.model.io cannot be resolved
        DependencyUpdatesReportXpp3Writer cannot be resolved to a type
        DependencyUpdatesReport cannot be resolved to a type
        DependencyReportSummary cannot be resolved to a type
        The method createDependencyInfo(Map<Dependency,ArtifactVersions>, boolean) from the type DependencyUpdatesXmlReportRenderer refers to the missing type DependencyInfo
        The method createDependencyInfo(Map<Dependency,ArtifactVersions>, boolean) from the type DependencyUpdatesXmlReportRenderer refers to the missing type DependencyInfo
        DependencyInfo cannot be resolved to a type
        DependencyInfo cannot be resolved to a type
        The method ifPresent(Consumer<? super String>) in the type Optional<String> is not applicable for the arguments (this::setLastVersion)
        The type new DependencyInfo(){} does not define setLastVersion(String) that is applicable here
        The method setSection(AbstractVersionDetails, Segment, Consumer<List<String>>, boolean) in the type CommonXmlReportRendererUtils is not applicable for the arguments (ArtifactVersions, Segment, this::setIncrementals, boolean)
        The type new DependencyInfo(){} does not define setIncrementals(List<String>) that is applicable here
        The method setSection(AbstractVersionDetails, Segment, Consumer<List<String>>, boolean) in the type CommonXmlReportRendererUtils is not applicable for the arguments (ArtifactVersions, Segment, this::setMinors, boolean)
        The type new DependencyInfo(){} does not define setMinors(List<String>) that is applicable here
        The method setSection(AbstractVersionDetails, Segment, Consumer<List<String>>, boolean) in the type CommonXmlReportRendererUtils is not applicable for the arguments (ArtifactVersions, Segment, this::setMajors, boolean)
        The type new DependencyInfo(){} does not define setMajors(List<String>) that is applicable here

        at org.codehaus.mojo.versions.xml.DependencyUpdatesXmlReportRenderer.<init>(DependencyUpdatesXmlReportRenderer.java:37)
        at org.codehaus.mojo.versions.DependencyUpdatesXmlRendererTest.testReportGeneration(DependencyUpdatesXmlRendererTest.java:89)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
```

jimisola commented:
> mvn verify -P run-its

Fixed.

slawekjaranowski commented:
On CI we don't have a git configuration for user

from build log:

```
Running post-build script: /home/runner/work/versions/versions/versions-maven-plugin/target/it/it-dynamic-versioning-scm-tag-latest-commit/verify.groovy
Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.
```

So pease set git user in verify scripts ... without `--global`, or add env `GIT_AUTHOR_NAME` , `GIT_AUTHOR_EMAIL` to invoker.properties
https://maven.apache.org/plugins/maven-invoker-plugin/run-mojo.html#invokerPropertiesFile

lfvjimisola commented:
I can't get it to work with neither environment variables in invoker.properties or git config commands in setup script (you wrote verify, but you meant setup right?). I could have understood if the environment variables were set in a different shell (scope) and not available later on but even when I set git config locally and see that they're set it's not picked up by git .

If is set it globally then it works. 

Any ideas? 

```
invoker.environmentVariables.GIT_AUTHOR_NAME=User Example
invoker.environmentVariables.GIT_AUTHOR_EMAIL=user@example.com
invoker.environmentVariables.GIT_DEFAULT_BRANCH=main
```

```groovy
[...]
exec('git config init.defaultBranch "main"')
exec('git config user.email "user@example.com"')
exec('git config user.name "User Example"')

exec('git config --list --local')

exec('git init')
exec('git add test.txt')
exec('git commit -m initial-commit')

[...]
```

As you can see below default branch, email and name are shown using `git config --list --local`, but git init does not pick it up and nor does `git commit...`. 

```
Running selector script: /home/u30576/dev/clones/github/jimisola/versions/versions-maven-plugin/target/it/it-dynamic-versioning-scm-tag-4-commits-1-vtag/selector.groovy
git version 2.45.2
Finished selector script: /home/u30576/dev/clones/github/jimisola/versions/versions-maven-plugin/target/it/it-dynamic-versioning-scm-tag-4-commits-1-vtag/selector.groovy
Running pre-build script: /home/u30576/dev/clones/github/jimisola/versions/versions-maven-plugin/target/it/it-dynamic-versioning-scm-tag-4-commits-1-vtag/setup.groovy
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
remote.origin.url=https://github.com/jimisola/versions.git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
remote.upstream.url=https://github.com/mojohaus/versions.git
remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
branch.feature/add_maven_ci_friendly_versions_goal.remote=origin
branch.feature/add_maven_ci_friendly_versions_goal.merge=refs/heads/feature/add_maven_ci_friendly_versions_goal
branch.feature/add_maven_ci_friendly_versions_goal.github-pr-owner-number=mojohaus#versions#1082
init.defaultbranch=main[user]
init.email=user@example.com
init.name=User Example
hint: Using 'master' as the name for the initial branch. This default branch name
Initialized empty Git repository in /home/u30576/dev/clones/github/jimisola/versions/versions-maven-plugin/target/it/it-dynamic-versioning-scm-tag-4-commits-1-vtag/.git/
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint:
hint: 	git config --global init.defaultBranch <name>
hint:
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint:
hint: 	git branch -m <name>
Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'name@hostname.(none)')
java.lang.AssertionError: Command 'git commit -m initial-commit' returned status: 128. Expression: (proc.exitValue() == 0)
```

lfvjimisola commented:
Running relevant git commands in setup.py as follows works:

`exec('git -c user.email="user@example.com" -c user.name="User" commit -m initial-commit')`

but it's really annoying having to submit it for each command.

Please advise if that is a solution we can use.

slawekjaranowski commented:
@lfvjimisola - ups ... my fault environment set in invoker.properties are available during test plugin execution
scripts like setup, verify are execute in context of current Maven session build 
so we need to add in setup.groovy

```
git config user.email "you@example.com"
```

lfvjimisola commented:
> @lfvjimisola - ups ... my fault environment set in invoker.properties are available during test plugin execution scripts like setup, verify are execute in context of current Maven session build so we need to add in setup.groovy
> 
> ```
> git config user.email "you@example.com"
> ```

I wrote above you reply that it does not work for me in setup.groovy with `exec('git config user.name "User Example"')`.

It works if:

1. -global is used with git config
2. user email and password is set explicitly with each command, i.e. `exec('git -c user.email="user@example.com" -c user.name="User" commit -m initial-commit')`

jimisola commented:
I just realized when I saw your commits that I might have executed git config prior to git init 🙄

slawekjaranowski commented on versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/DynamicVersioningSCMPlugin.java:
```diff
@@ -0,0 +1,227 @@
+@Mojo(name = "set-version-from-scm-tag", defaultPhase = LifecyclePhase.INITIALIZE)
+public class DynamicVersioningSCMPlugin extends AbstractMojo {
+
+    /**
+     * The Maven Project Object
+     */
+    @Parameter(defaultValue = "${project}", readonly = true, required = true)
```
We also should add `@since` tag to all parameters.

slawekjaranowski commented on versions-maven-plugin/src/it/it-dynamic-versioning-scm-tag-property-name/invoker.properties:
```diff
@@ -0,0 +1,4 @@
+invoker.goals=${project.groupId}:${project.artifactId}:${project.version}:set-version-from-scm-tag -DpropertyName=version_from_scm
+invoker.environmentVariables.GIT_AUTHOR_NAME=User Example
+invoker.environmentVariables.GIT_AUTHOR_EMAIL=user@example.com
+invoker.environmentVariables.GIT_DEFAULT_BRANCH=main
```
env can be removed

slawekjaranowski reviewed (commented):
We should add goal to list at page: https://www.mojohaus.org/versions/versions-maven-plugin/index.html

We can also add en example page with examples - what version will be set according to git repo content.

slawekjaranowski commented:
ok, looks good enough

open questions:
 - goal name
 - documentation - for it I will accept a separate PR

jimisola commented:
* Renamed goal to `use-dynamic-version-from-scm`
* Added docs

FYI, there is a discrepancy between the goals listed in

https://www.mojohaus.org/versions/versions-maven-plugin/#goals-overview and https://www.mojohaus.org/versions/versions-maven-plugin/plugin-info.html

Is there a reason why Single Source of Truth (SSOT) and only the auto-generated list is used?
I understand that there a normal goals and reporting goals, but does there have to be a manually generated list to list them? Wouldn't it be enough to just refer to Goals page and clarify that there are different types of goals ("normal" and reporting)?

slawekjaranowski commented:
> 
> Is there a reason why Single Source of Truth (SSOT) and only the auto-generated list is used? I understand that there a normal goals and reporting goals, but does there have to be a manually generated list to list them? Wouldn't it be enough to just refer to Goals page and clarify that there are different types of goals ("normal" and reporting)?

You right you can think about to drop manual docs at all in order to use generated one.

lfvjimisola commented:
Thank you @slawekjaranowski

slawekjaranowski commented:
@lfvjimisola thanks also for contributing - I will release next version in a few days
